definitions/npm/shelljs_v0.7.x/flow_>=v0.28.x/shelljs_v0.7.x.js

```diff
@@ -0,0 +1,381 @@
+  & ((pattern: $npm$shelljs$Pattern, subst: string, glob: string, ...rest: string[]) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$CmdSet =
+  & ((exitOnError: '-e' | '+e', rest: void) => void)
+  & ((verbose: '-v' | '+v', rest: void) => void)
+  & ((disableGlobbing: '-f' | '+f', rest: void) => void)
+  ;
+
+declare type $npm$shelljs$OptSort = $npm$shelljs$ICmdOptions<'-n' | '-r'>;
+declare type $npm$shelljs$CmdSort =
+  & ((options: $npm$shelljs$OptSort, glob: string, ...rest: string[]) => $npm$shelljs$String)
+  & ((glob: string, ...rest: string[]) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$CmdTail =
+  & ((num: number, glob: string, ...rest: string[]) => $npm$shelljs$String)
+  & ((glob: string, ...rest: string[]) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$CmdTempdir =
+  & ((rest: void) => string)
+  ;
+
+declare type $npm$shelljs$OptTest = '-b' | '-c' | '-d' | '-e' | '-f' | '-L' | '-p' | '-S';
+declare type $npm$shelljs$CmdTest =
+  & ((mode: $npm$shelljs$OptTest, path: string, rest: void) => boolean)
+  ;
+
+declare type $npm$shelljs$OptTouch = {
+  [key: '-a' | '-c' | '-m' ]: true;
+  '-d'?: string;
+  '-r'?: string;
+};
+declare type $npm$shelljs$CmdTouch =
+  & ((options: $npm$shelljs$OptTouch, glob: string, ...rest: string[]) => $npm$shelljs$String)
+  & ((glob: string, ...rest: string[]) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$CmdWhich =
+  & ((cmd: string, rest: void) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$PipeCat =
+  & ((rest: void) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$PipeExec =
+  & ((cmd: string, options: $npm$shelljs$OptExec & { async: true }, callback: $npm$shelljs$FunExec, rest: void) => $npm$shelljs$Async)
+  & ((cmd: string, options: $npm$shelljs$OptExec & { async: true }, rest: void) => $npm$shelljs$Async)
+  & ((cmd: string, options: $npm$shelljs$OptExecSync, rest: void) => $npm$shelljs$String)
+  & ((cmd: string, rest: void) => $npm$shelljs$String)
+  & ((cmd: string, callback: $npm$shelljs$FunExec, rest: void) => $npm$shelljs$Async)
+  ;
+
+declare type $npm$shelljs$PipeGrep =
+  & ((options: $npm$shelljs$OptGrep, pattern: $npm$shelljs$Pattern, rest: void) => $npm$shelljs$String)
+  & ((pattern: $npm$shelljs$Pattern, rest: void) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$PipeHead =
+  & ((num: number, rest: void) => $npm$shelljs$String)
+  & ((rest: void) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$PipeSed =
+  & ((pattern: $npm$shelljs$Pattern, subst: string, rest: void) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$PipeSort =
+  & ((options: $npm$shelljs$OptSort, rest: void) => $npm$shelljs$String)
+  & ((rest: void) => $npm$shelljs$String)
+  ;
+
+declare type $npm$shelljs$PipeTail =
+  & ((num: number, rest: void) => $npm$shelljs$String)
+  & ((rest: void) => $npm$shelljs$String)
+  ;
+
+declare module 'shelljs' {
```

Comment:
Interesting. I just experimented with `declare export` and I think I see what you mean. I should be able to clean some of this up that way.

Regarding namespaced globals, the main reason I have been using them in my libdefs so far is so that I can refer to types defined in libdefs for some module `m0` in libdefs for another module `m1`.

For instance, I have another small library built on `shelljs` that provides a promise based API for some parts like `exec`. I use flow for the code and babel + rollup to transform it to normal javascript on prepublish. Along with the package, I provide a libdef that refers to the types defined in this one.  In order to make the libdef for my library usable, it seemed like I needed to use namespaced globals for the `shelljs` libdef.

Is there another way to preserve this type information across module/package boundaries? I've been keeping an eye on https://github.com/flowtype/flow-typed/issues/16 since it seems relevant.

Just to be clear, here's a simplified example of the scenario I'm describing:

```javascript
// .flowconfig
[libs]
libdef-m0.js
libdef-m1.js

// libdef-m0.js
declare module "m0" {
  declare interface I0 {
    p0: T0;
  };
}

// libdef-m1.js
declare module "m1" {
  declare interface I1 {
    // How do we provide the type for p0? We can `import type { … }` in this
    // file (libdef-m1.js) but flow doesn't seem to preserve that information
    // when we use this as a lib from another file (see below).
    p0: m0.I0;
    p1: string;
  };
}

// index.js
// @flow
import type { I1 } from "m1";

declare var foo: I1;
// foo.p0 <- flow reports no type
// foo.p1 <- flow reports type string
```